VMSS Flex Support: VMSet implementation part 4

zmyzheng opened the pull request:
VMSS Flex Support: VMSet implementation part 4

<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?
/kind feature
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature
/kind design

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:
This PR is the part 4 of FlexScaleSet implementation of VMSet

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Refer to #639 

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

bridgetkromhout commented:
/ok-to-test

zmyzheng commented:
/retest

nilo19 commented:
The test coverage decreased. Can we enrich the unit test?

zmyzheng commented:
> The test coverage decreased. Can we enrich the unit test?

Not sure how the test coverage is calculated, but I have added unit tests for each newly added functions in this PR.

nilo19 commented:
> > The test coverage decreased. Can we enrich the unit test?
> 
> Not sure how the test coverage is calculated, but I have added unit tests for each newly added functions in this PR.

Have you covered all new lines?

zmyzheng commented:
> > > The test coverage decreased. Can we enrich the unit test?
> > 
> > 
> > Not sure how the test coverage is calculated, but I have added unit tests for each newly added functions in this PR.
> 
> Have you covered all new lines?

I see. I misunderstood how the test coverage is calculated. I just added more test cases to handle differerent scenarios. Please take a look again.

nilo19 commented:
/lgtm

nilo19 commented on pkg/provider/azure_vmssflex_cache_test.go:
```diff
@@ -103,8 +105,43 @@ var (
+		Location: to.StringPtr("EastUS"),
+	}
+
+	testVMWithoutInstanceView3 = compute.VirtualMachine{
```
Can we wrap these into a function to reduce code redundancy?

zmyzheng replied:
Sounds good. Just added a helper function to wrap them

nilo19 commented:
/lgtm
/approve

zmyzheng commented:
/retest
